### PR TITLE
Synchronize VM template labels to VMI

### PR DESCRIPTION
**Pull request by yaocw2020**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**

We have to restart VM to synchronize the labels to VMI. It will cause request interruption.

**Solution:**

Add a VM controller to synchronize the VM template labels to the relevant VMI.

**Related Issue:**
https://github.com/harvester/harvester/issues/1676

**Test plan:**

### CODE REVIEW

File: pkg/controller/master/virtualmachine/vm_controller.go [hunk 1]
```diff
@@ -142,6 +145,50 @@ func (h *VMController) SetOwnerOfPVCs(_ string, vm *kubevirtv1.VirtualMachine) (
...
+	}
+
+	vmi, err := h.vmiCache.Get(vm.Namespace, vm.Name)
+	if err != nil && apierrors.IsNotFound(err) {
+		return vm, nil
+	} else if err != nil {
+		return nil, fmt.Errorf("get vmi %s/%s failed, error: %w", vm.Namespace, vm.Name, err)
+	}
+
+	var vmiCopy *kubevirtv1.VirtualMachineInstance
```
Comment (guangbochen) [inline pkg/controller/master/virtualmachine/vm_controller.go]:
```suggestion
	vmiCopy := vmi.DeepCopy()
```

File: pkg/controller/master/virtualmachine/vm_controller.go [hunk 2]
```diff
@@ -142,6 +145,50 @@ func (h *VMController) SetOwnerOfPVCs(_ string, vm *kubevirtv1.VirtualMachine) (
...
+	for k := range vmi.Labels {
+		if _, ok := vm.Spec.Template.ObjectMeta.Labels[k]; !strings.HasPrefix(k, kubevirt.GroupName) && !ok {
+			if vmiCopy == nil {
+				vmiCopy = vmi.DeepCopy()
+			}
+			delete(vmiCopy.Labels, k)
+		}
+	}
+
+	if vmiCopy != nil {
```
Comment (guangbochen) [inline pkg/controller/master/virtualmachine/vm_controller.go]: u can compare the vmi and vmiCopy obj instead.

File: pkg/controller/master/virtualmachine/vm_controller.go [hunk 3]
```diff
@@ -142,6 +145,50 @@ func (h *VMController) SetOwnerOfPVCs(_ string, vm *kubevirtv1.VirtualMachine) (
 	return vm, nil
 }
 
+// SyncLabelsToVmi synchronizes the labels in the VM template to the existed VMI
```
Comment (guangbochen) [inline pkg/controller/master/virtualmachine/vm_controller.go]:
```suggestion
// SyncLabelsToVmi synchronizes the labels in the VM spec to the existing VMI without re-deployment.
```

**Review by weihanglo — COMMENTED** [review]
The code looks fine to me. Just wonder why #1676 need this feature. Could you elaborate the relevance in PR description? That helps others traceback in the future. Thanks!

File: pkg/controller/master/virtualmachine/vm_controller.go [hunk 4]
```diff
@@ -142,6 +146,55 @@ func (h *VMController) SetOwnerOfPVCs(_ string, vm *kubevirtv1.VirtualMachine) (
...
+	// Add or update the labels of VMI
+	for k, v := range vm.Spec.Template.ObjectMeta.Labels {
+		if !strings.HasPrefix(k, kubevirt.GroupName) && vmi.Labels[k] != v {
+			if len(vmiCopy.Labels) == 0 {
+				vmiCopy.Labels = make(map[string]string)
+			}
+			vmiCopy.Labels[k] = v
+		}
+	}
+	// delete the labels existing in the vmi.Labels but not in the vm.spec.template.objectMeta.Labels
```
Comment (guangbochen) [inline pkg/controller/master/virtualmachine/vm_controller.go]:
```suggestion
	// delete the labels exist in the `vmi.Labels` but not in the `vm.spec.template.objectMeta.Labels`
```

**Review by futuretea — APPROVED** [review]
LGTM

File: pkg/controller/master/virtualmachine/vm_controller.go [hunk 5]
```diff
@@ -142,6 +146,55 @@ func (h *VMController) SetOwnerOfPVCs(_ string, vm *kubevirtv1.VirtualMachine) (
...
+	for k := range vmi.Labels {
+		if _, ok := vm.Spec.Template.ObjectMeta.Labels[k]; !strings.HasPrefix(k, kubevirt.GroupName) && !ok {
+			delete(vmiCopy.Labels, k)
+		}
```
Comment (guangbochen) [inline pkg/controller/master/virtualmachine/vm_controller.go:186]: Does removing the kubevirt-specific labels will cause any side effect?

Comment (yaocw2020) [reply pkg/controller/master/virtualmachine/vm_controller.go:186]: Kubevirt has a webhook to avoid those labels from deleting.

**Review by guangbochen — APPROVED** [review]
LGTM

**Pull request merged**